File: backend/app/schemas/shipping.py
"""Shipping / serviceability schemas."""
from pydantic import BaseModel, Field


class ServiceabilityRequest(BaseModel):
    pincode: str = Field(min_length=4, max_length=10)
    weight_grams: int = Field(default=500, ge=1)


class CourierOption(BaseModel):
    courier_name: str | None = None
    courier_id: int | None = None
    rate: float
    eta_days: int | str | None = None


class ServiceabilityResponse(BaseModel):
    serviceable: bool
    couriers: list[CourierOption] = []
    cheapest: CourierOption | None = None

File: frontend/src/api/shipping.ts
import { apiClient } from "@/api/client";

export interface CourierOption {
  courier_name: string | null;
  courier_id: number | null;
  rate: number;
  eta_days: number | string | null;
}

export interface Serviceability {
  serviceable: boolean;
  couriers: CourierOption[];
  cheapest: CourierOption | null;
}

export async function checkServiceability(
  pincode: string,
  weightGrams = 500,
): Promise<Serviceability> {
  const { data } = await apiClient.post<Serviceability>("/shipping/check-serviceability", {
    pincode,
    weight_grams: weightGrams,
  });
  return data;
}
